src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentStepExecution.java

```diff
@@ -0,0 +1,128 @@
+package com.atlassian.bitbucket.jenkins.internal.deployments;
+
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.DeploymentState;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepository;
+import com.atlassian.bitbucket.jenkins.internal.status.BitbucketRevisionAction;
+import com.google.common.annotations.VisibleForTesting;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepExecutionImpl;
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+
+import java.util.logging.Logger;
+
+public class DeploymentStepExecution extends AbstractStepExecutionImpl {
+
+    private static final long serialVersionUID = 1L;
+
+    private final transient BitbucketDeploymentFactory deploymentFactory;
+    private final transient DeploymentPoster deploymentPoster;
+    private final BitbucketDeploymentEnvironment environment;
+
+    DeploymentStepExecution(BitbucketDeploymentEnvironment environment,
+                            StepContext context) {
+        this(environment, context, getStepDescriptor().getBitbucketDeploymentFactory(), getStepDescriptor().getDeploymentPoster());
+    }
+
+    @VisibleForTesting
+    DeploymentStepExecution(BitbucketDeploymentEnvironment environment,
+                            StepContext context,
+                            BitbucketDeploymentFactory deploymentFactory,
+                            DeploymentPoster deploymentPoster) {
+        super(context);
+        this.environment = environment;
+        this.deploymentPoster = deploymentPoster;
+        this.deploymentFactory = deploymentFactory;
+    }
+
+    @Override
+    public boolean start() throws Exception {
+        getContext().newBodyInvoker()
+                .withCallback(new Callback(environment, deploymentFactory, deploymentPoster))
+                .start();
+        return false;
+    }
+
+    @VisibleForTesting
+    BitbucketDeploymentEnvironment getEnvironment() {
+        return environment;
+    }
+
+    private static DeploymentStepImpl.DescriptorImpl getStepDescriptor() {
```

Comment:
As above, I would make this public. It can be static but non-static makes testing easier, and a consumer that needs a static callout can always do a `Jenkins.getDescriptorByType` call themselves.